File: apps/ios/CoreHub/Phone/DeviceProof.swift
// The push relay's device proof (ADR 0024 §6, the contract's `PushRelayProof`). A hub with no
// APNs key of its own pushes through the Core Hub relay, which binds this phone's token to the
// first hub that asks. When the phone signs in to another hub without signing out of the first,
// this proof moves the token at once instead of after 30 days: a P-256 key made once for this
// install, kept in the Keychain, signs `corehub-push-bind-v1`, the platform, the token and the
// time; the hub forwards it to the relay unread. The relay moves a binding only for a newer
// proof from the key it recorded first.
import CoreHubClient
import CryptoKit
import Foundation
import Security

/// Where this install's proof key lives.
protocol ProofKeyStore {
    func read() -> Data?
    /// True when the key is kept: a key that is not kept proves nothing the next time.
    func write(_ data: Data) -> Bool
}

/// The Keychain, on this device only (never in a backup that lands on another phone), readable
/// after the first unlock since a new APNs token may arrive with the phone locked.
struct KeychainProofKeyStore: ProofKeyStore {
    var service = Product.storagePrefix + "push-proof"
    var account = "p256"

    func read() -> Data? {
        var query = baseQuery
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else { return nil }
        return item as? Data
    }

    func write(_ data: Data) -> Bool {
        var add = baseQuery
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        if status == errSecDuplicateItem {
            let update: [String: Any] = [kSecValueData as String: data]
            return SecItemUpdate(baseQuery as CFDictionary, update as CFDictionary) == errSecSuccess
        }
        return status == errSecSuccess
    }

    private var baseQuery: [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
    }
}

struct DeviceProof {
    /// The relay's prefix (packages/push-relay/src/relay.ts `PROOF_PREFIX`).
    static let prefix = "corehub-push-bind-v1"

    let store: ProofKeyStore

    /// This install's key: made the first time it is needed, the same one every time after.
    func key() -> P256.Signing.PrivateKey? {
        if let data = store.read(), let kept = try? P256.Signing.PrivateKey(rawRepresentation: data) {
            return kept
        }
        let fresh = P256.Signing.PrivateKey()
        return store.write(fresh.rawRepresentation) ? fresh : nil
    }

    /// What is signed: the prefix, the platform (`apns`), the token and the unix time, one a line.
    static func message(platform: String, token: String, signedAt: Int) -> Data {
        Data([prefix, platform, token, String(signedAt)].joined(separator: "\n").utf8)
    }

    /// The proof for one registration: the raw public point (65 bytes) and the raw `r || s`
    /// signature (64 bytes), both base64url. Nil when the key cannot be made or kept; the phone
    /// then registers without one, as an older app does.
    func proof(platform: String, token: String, now: Date = Date()) -> PushRelayProof? {
        guard let key = key() else { return nil }
        let signedAt = Int(now.timeIntervalSince1970)
        let message = Self.message(platform: platform, token: token, signedAt: signedAt)
        guard let signature = try? key.signature(for: message) else { return nil }
        return PushRelayProof(
            key: Self.base64url(key.publicKey.x963Representation),
            signedAt: signedAt,
            signature: Self.base64url(signature.rawRepresentation)
        )
    }

    static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: apps/android/app/src/main/java/hub/core/android/phone/DeviceProof.kt
package hub.core.android.phone

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import hub.core.client.model.PushRelayProof
import java.math.BigInteger
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.util.Base64

/**
 * The push relay's device proof (ADR 0024 §6, the contract's `PushRelayProof`). A hub with no FCM
 * key of its own pushes through the Core Hub relay, which binds this phone's token to the first
 * hub that asks. When the phone signs in to another hub without signing out of the first, this
 * proof moves the token at once instead of after 30 days: a P-256 key made once for this install,
 * kept in the Android Keystore (its private half never leaves it), signs `corehub-push-bind-v1`,
 * the platform, the token and the time; the hub forwards it to the relay unread. The relay moves a
 * binding only for a newer proof from the key it recorded first.
 */
fun interface ProofKeys {
    /** This install's key pair, made the first time and the same after; null when it cannot be. */
    fun keyPair(): KeyPair?
}

/** The Android Keystore: the key is made there and only used there. */
class KeystoreProofKeys(private val alias: String = "corehub.push-proof") : ProofKeys {
    override fun keyPair(): KeyPair? = runCatching {
        val store = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        val kept = store.getEntry(alias, null) as? KeyStore.PrivateKeyEntry
        if (kept != null) {
            KeyPair(kept.certificate.publicKey, kept.privateKey)
        } else {
            KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE).run {
                initialize(
                    KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_SIGN)
                        .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                        .setDigests(KeyProperties.DIGEST_SHA256)
                        .build(),
                )
                generateKeyPair()
            }
        }
    }.getOrNull()

    private companion object {
        const val KEYSTORE = "AndroidKeyStore"
    }
}

object DeviceProof {
    /** The relay's prefix (packages/push-relay/src/relay.ts `PROOF_PREFIX`). */
    const val PREFIX = "corehub-push-bind-v1"

    /** What is signed: the prefix, the platform (`fcm`), the token and the unix time, one a line. */
    fun message(platform: String, token: String, signedAt: Long): ByteArray =
        listOf(PREFIX, platform, token, signedAt.toString()).joinToString("\n").toByteArray(Charsets.UTF_8)

    /**
     * The proof for one registration: the raw public point (65 bytes) and the raw `r || s`
     * signature (64 bytes), both base64url. Null when the key cannot be made or used; the phone
     * then registers without one, as an older app does.
     */
    fun proof(keys: ProofKeys, platform: String, token: String, nowMillis: Long = System.currentTimeMillis()): PushRelayProof? =
        runCatching {
            val pair = keys.keyPair() ?: return null
            val public = pair.public as? ECPublicKey ?: return null
            val signedAt = nowMillis / 1000
            val der = Signature.getInstance("SHA256withECDSA").run {
                initSign(pair.private)
                update(message(platform, token, signedAt))
                sign()
            }
            PushRelayProof(
                key = base64url(rawPublicKey(public)),
                signedAt = signedAt.toInt(),
                signature = base64url(rawSignature(der)),
            )
        }.getOrNull()

    /** The uncompressed point `04 || x || y`, each coordinate 32 bytes. */
    fun rawPublicKey(key: ECPublicKey): ByteArray =
        byteArrayOf(4) + fixed(key.w.affineX) + fixed(key.w.affineY)

    /** Java signs ECDSA as DER `SEQUENCE { INTEGER r, INTEGER s }`; the relay reads raw `r || s`. */
    fun rawSignature(der: ByteArray): ByteArray {
        var at = 0
        fun length(): Int {
            val first = der[at++].toInt() and 0xff
            if (first < 0x80) return first
            var value = 0
            repeat(first and 0x7f) { value = (value shl 8) or (der[at++].toInt() and 0xff) }
            return value
        }
        fun integer(): BigInteger {
            require(der[at++].toInt() == 0x02) { "not a DER integer" }
            val size = length()
            return BigInteger(1, der.copyOfRange(at, at + size)).also { at += size }
        }
        require(der[at++].toInt() == 0x30) { "not a DER sequence" }
        length()
        return fixed(integer()) + fixed(integer())
    }

    fun base64url(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    private fun fixed(value: BigInteger): ByteArray {
        val bytes = value.toByteArray().dropWhile { it == 0.toByte() }.toByteArray()
        require(bytes.size <= 32) { "a P-256 value is 32 bytes" }
        return ByteArray(32 - bytes.size) + bytes
    }
}
